numpy/f2py/src/fortranobject.c

```diff
@@ -690,32 +709,119 @@ static int swap_arrays(PyArrayObject* obj1, PyArrayObject* obj2) {
        ||(PyArray_ISFLOAT(arr) && PyTypeNum_ISFLOAT(type_num))          \
        ||(PyArray_ISCOMPLEX(arr) && PyTypeNum_ISCOMPLEX(type_num))      \
        ||(PyArray_ISBOOL(arr) && PyTypeNum_ISBOOL(type_num))            \
+       ||(PyArray_ISSTRING(arr) && PyTypeNum_ISSTRING(type_num))        \
        )
 
-extern
-PyArrayObject* array_from_pyobj(const int type_num,
-                                npy_intp *dims,
-                                const int rank,
-                                const int intent,
-                                PyObject *obj) {
+/*
+  get_elsize determines array itemsize from a Python object.  Returns
+  elsize if succesful, -1 otherwise.
+
+  Supported types of the input are: numpy.ndarray, bytes, str, tuple,
+  list.
+ */
+static int
+get_elsize(PyObject *obj) {
+  if (PyArray_Check(obj)) {
+    return PyArray_DESCR((PyArrayObject *)obj)->elsize;
+  } else if (PyBytes_Check(obj)) {
+    return PyBytes_GET_SIZE(obj);
+  } else if (PyUnicode_Check(obj)) {
+    return PyUnicode_GET_LENGTH(obj);
+  } else if (PySequence_Check(obj)) {
+    PyObject* fast = PySequence_Fast(obj, "f2py:fortranobject.c:get_elsize");
+    if (fast != NULL) {
+      Py_ssize_t i, n = PySequence_Fast_GET_SIZE(fast);
+      int sz, elsize = 0;
+      for (i=0; i<n; i++) {
+        sz = get_elsize(PySequence_Fast_GET_ITEM(fast, i) /* borrowed */);
+        if (sz > elsize) {
+          elsize = sz;
+        }
+      }
+      Py_DECREF(fast);
+      return elsize;
+    }
+  }
+  return -1;
+}
+
+extern PyArrayObject *
+ndarray_from_pyobj(const int type_num,
+                   const int elsize_,
+                   npy_intp *dims,
+                   const int rank,
+                   const int intent,
+                   PyObject *obj,
+                   const char *errmess) {
     /*
+     * Return an array with given element type and shape from a Python
+     * object while taking into account the usage intent of the array.
+     *
+     * - element type is defined by type_num and elsize
+     * - shape is defined by dims and rank
+     *
+     * ndarray_from_pyobj is used to convert Python object arguments
+     * to numpy ndarrays with given type and shape that data is passed
+     * to interfaced Fortran or C functions.
+     *
+     * errmess (if not NULL), contains a prefix of an error message
+     * for an exception to be triggered within this function.
+     *
+     * Negative elsize value means that elsize is to be determined
+     * from the Python object in runtime.
+     *
+     * Note on strings
```

Comment:
Users of f2py are not affected by this note as they would not use `ndarray_from_pyobj` directly, not even in the advanced usage case. It would make sense to add these notes to f2py internal/developer documentation that we currently have in the form of source code.